File: 25-comboBreaker/go/challenge/partOne.go
package challenge

import "errors"

func singleTransformation(v, subjectNumber int) int {
	return (v * subjectNumber) % 20201227
}

func transform(loopSize, subjectNumber int) int {
	v := 1
	for i := 0; i < loopSize; i += 1 {
		v = singleTransformation(v, subjectNumber)
	}
	return v
}

func findLoopSize(targetPubkey int) int {
	var c int
	v := 1
	for {
		c += 1
		v = singleTransformation(v, 7)
		if v == targetPubkey {
			return c
		}
	}
}

func PartOne(instr string) int {
	inputSlice := parse(instr)

	pubkeyOne := inputSlice[0]
	pubkeyTwo := inputSlice[1]

	loopSizeOne := findLoopSize(pubkeyOne)
	loopSizeTwo := findLoopSize(pubkeyTwo)

	encryptionKey := transform(loopSizeTwo, pubkeyOne)
	if encryptionKey != transform(loopSizeOne, pubkeyTwo) {
		panic(errors.New("encryption keys do not match"))
	}

	return encryptionKey
}

File: challenges/2020/25-comboBreaker/python/partOne.py
from common import *


def single_transformation(v:int, subject_number:int=7) -> int:
    return (v * subject_number) % 20201227


def transform(loop_size:int, subject_number:int=7) -> int:
    v = 1
    for _ in range(loop_size):
        v = single_transformation(v, subject_number=subject_number)
    return v


def find_loop_size(target_pubkey:int) -> int:
    c = 0
    v = 1
    while True:
        c += 1
        v = single_transformation(v)
        if v == target_pubkey:
            return c


def partOne(instr: str) -> int:
    pubkey_one, pubkey_two = parse(instr)

    loop_size_one = find_loop_size(pubkey_one)
    loop_size_two = find_loop_size(pubkey_two)

    encryption_key = transform(loop_size_two, subject_number=pubkey_one)
    assert encryption_key == transform(loop_size_one, subject_number=pubkey_two), "encryption keys do not match"

    return encryption_key
